{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "import ray\n",
    "import modules.network_extractor as net_extractor\n",
    "import geopandas as geopd\n",
    "import pandas as pd\n",
    "from shapely import ops\n",
    "from osmnx import settings\n",
    "import pyproj\n",
    "from pathlib import Path\n",
    "import networkx as nx\n",
    "import osmnx as ox"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-04-30 12:02:33,578\tINFO worker.py:1832 -- Started a local Ray instance. View the dashboard at \u001b[1m\u001b[32mhttp://127.0.0.1:8265 \u001b[39m\u001b[22m\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "14a4ced146e44c27a1d112209f2fac26",
       "version_major": 2,
       "version_minor": 0
      },
      "text/html": [
       "<div class=\"lm-Widget p-Widget lm-Panel p-Panel jp-Cell-outputWrapper\">\n",
       "    <div style=\"margin-left: 50px;display: flex;flex-direction: row;align-items: center\">\n",
       "        <div class=\"jp-RenderedHTMLCommon\" style=\"display: flex; flex-direction: row;\">\n",
       "  <svg viewBox=\"0 0 567 224\" fill=\"none\" xmlns=\"http://www.w3.org/2000/svg\" style=\"height: 3em;\">\n",
       "    <g clip-path=\"url(#clip0_4338_178347)\">\n",
       "        <path d=\"M341.29 165.561H355.29L330.13 129.051C345.63 123.991 354.21 112.051 354.21 94.2307C354.21 71.3707 338.72 58.1807 311.88 58.1807H271V165.561H283.27V131.661H311.8C314.25 131.661 316.71 131.501 319.01 131.351L341.25 165.561H341.29ZM283.29 119.851V70.0007H311.82C331.3 70.0007 342.34 78.2907 342.34 94.5507C342.34 111.271 331.34 119.861 311.82 119.861L283.29 119.851ZM451.4 138.411L463.4 165.561H476.74L428.74 58.1807H416L367.83 165.561H380.83L392.83 138.411H451.4ZM446.19 126.601H398L422 72.1407L446.24 126.601H446.19ZM526.11 128.741L566.91 58.1807H554.35L519.99 114.181L485.17 58.1807H472.44L514.01 129.181V165.541H526.13V128.741H526.11Z\" fill=\"var(--jp-ui-font-color0)\"/>\n",
       "        <path d=\"M82.35 104.44C84.0187 97.8827 87.8248 92.0678 93.1671 87.9146C98.5094 83.7614 105.083 81.5067 111.85 81.5067C118.617 81.5067 125.191 83.7614 130.533 87.9146C135.875 92.0678 139.681 97.8827 141.35 104.44H163.75C164.476 101.562 165.622 98.8057 167.15 96.2605L127.45 56.5605C121.071 60.3522 113.526 61.6823 106.235 60.3005C98.9443 58.9187 92.4094 54.9203 87.8602 49.0574C83.3109 43.1946 81.0609 35.8714 81.5332 28.4656C82.0056 21.0599 85.1679 14.0819 90.4252 8.8446C95.6824 3.60726 102.672 0.471508 110.08 0.0272655C117.487 -0.416977 124.802 1.86091 130.647 6.4324C136.493 11.0039 140.467 17.5539 141.821 24.8501C143.175 32.1463 141.816 39.6859 138 46.0505L177.69 85.7505C182.31 82.9877 187.58 81.4995 192.962 81.4375C198.345 81.3755 203.648 82.742 208.33 85.3976C213.012 88.0532 216.907 91.9029 219.616 96.5544C222.326 101.206 223.753 106.492 223.753 111.875C223.753 117.258 222.326 122.545 219.616 127.197C216.907 131.848 213.012 135.698 208.33 138.353C203.648 141.009 198.345 142.375 192.962 142.313C187.58 142.251 182.31 140.763 177.69 138L138 177.7C141.808 184.071 143.155 191.614 141.79 198.91C140.424 206.205 136.44 212.75 130.585 217.313C124.731 221.875 117.412 224.141 110.004 223.683C102.596 223.226 95.6103 220.077 90.3621 214.828C85.1139 209.58 81.9647 202.595 81.5072 195.187C81.0497 187.779 83.3154 180.459 87.878 174.605C92.4405 168.751 98.9853 164.766 106.281 163.401C113.576 162.035 121.119 163.383 127.49 167.19L167.19 127.49C165.664 124.941 164.518 122.182 163.79 119.3H141.39C139.721 125.858 135.915 131.673 130.573 135.826C125.231 139.98 118.657 142.234 111.89 142.234C105.123 142.234 98.5494 139.98 93.2071 135.826C87.8648 131.673 84.0587 125.858 82.39 119.3H60C58.1878 126.495 53.8086 132.78 47.6863 136.971C41.5641 141.163 34.1211 142.972 26.7579 142.059C19.3947 141.146 12.6191 137.574 7.70605 132.014C2.79302 126.454 0.0813599 119.29 0.0813599 111.87C0.0813599 104.451 2.79302 97.2871 7.70605 91.7272C12.6191 86.1673 19.3947 82.5947 26.7579 81.6817C34.1211 80.7686 41.5641 82.5781 47.6863 86.7696C53.8086 90.9611 58.1878 97.2456 60 104.44H82.35ZM100.86 204.32C103.407 206.868 106.759 208.453 110.345 208.806C113.93 209.159 117.527 208.258 120.522 206.256C123.517 204.254 125.725 201.276 126.771 197.828C127.816 194.38 127.633 190.677 126.253 187.349C124.874 184.021 122.383 181.274 119.205 179.577C116.027 177.88 112.359 177.337 108.826 178.042C105.293 178.746 102.113 180.654 99.8291 183.44C97.5451 186.226 96.2979 189.718 96.3 193.32C96.2985 195.364 96.7006 197.388 97.4831 199.275C98.2656 201.163 99.4132 202.877 100.86 204.32ZM204.32 122.88C206.868 120.333 208.453 116.981 208.806 113.396C209.159 109.811 208.258 106.214 206.256 103.219C204.254 100.223 201.275 98.0151 197.827 96.97C194.38 95.9249 190.676 96.1077 187.348 97.4873C184.02 98.8669 181.274 101.358 179.577 104.536C177.879 107.714 177.337 111.382 178.041 114.915C178.746 118.448 180.653 121.627 183.439 123.911C186.226 126.195 189.717 127.443 193.32 127.44C195.364 127.443 197.388 127.042 199.275 126.259C201.163 125.476 202.878 124.328 204.32 122.88ZM122.88 19.4205C120.333 16.8729 116.981 15.2876 113.395 14.9347C109.81 14.5817 106.213 15.483 103.218 17.4849C100.223 19.4868 98.0146 22.4654 96.9696 25.9131C95.9245 29.3608 96.1073 33.0642 97.4869 36.3922C98.8665 39.7202 101.358 42.4668 104.535 44.1639C107.713 45.861 111.381 46.4036 114.914 45.6992C118.447 44.9949 121.627 43.0871 123.911 40.301C126.195 37.515 127.442 34.0231 127.44 30.4205C127.44 28.3772 127.038 26.3539 126.255 24.4664C125.473 22.5788 124.326 20.8642 122.88 19.4205ZM19.42 100.86C16.8725 103.408 15.2872 106.76 14.9342 110.345C14.5813 113.93 15.4826 117.527 17.4844 120.522C19.4863 123.518 22.4649 125.726 25.9127 126.771C29.3604 127.816 33.0638 127.633 36.3918 126.254C39.7198 124.874 42.4664 122.383 44.1635 119.205C45.8606 116.027 46.4032 112.359 45.6988 108.826C44.9944 105.293 43.0866 102.114 40.3006 99.8296C37.5145 97.5455 34.0227 96.2983 30.42 96.3005C26.2938 96.3018 22.337 97.9421 19.42 100.86ZM100.86 100.86C98.3125 103.408 96.7272 106.76 96.3742 110.345C96.0213 113.93 96.9226 117.527 98.9244 120.522C100.926 123.518 103.905 125.726 107.353 126.771C110.8 127.816 114.504 127.633 117.832 126.254C121.16 124.874 123.906 122.383 125.604 119.205C127.301 116.027 127.843 112.359 127.139 108.826C126.434 105.293 124.527 102.114 121.741 99.8296C118.955 97.5455 115.463 96.2983 111.86 96.3005C109.817 96.299 107.793 96.701 105.905 97.4835C104.018 98.2661 102.303 99.4136 100.86 100.86Z\" fill=\"#00AEEF\"/>\n",
       "    </g>\n",
       "    <defs>\n",
       "        <clipPath id=\"clip0_4338_178347\">\n",
       "            <rect width=\"566.93\" height=\"223.75\" fill=\"white\"/>\n",
       "        </clipPath>\n",
       "    </defs>\n",
       "  </svg>\n",
       "</div>\n",
       "\n",
       "        <table class=\"jp-RenderedHTMLCommon\" style=\"border-collapse: collapse;color: var(--jp-ui-font-color1);font-size: var(--jp-ui-font-size1);\">\n",
       "    <tr>\n",
       "        <td style=\"text-align: left\"><b>Python version:</b></td>\n",
       "        <td style=\"text-align: left\"><b>3.11.3</b></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td style=\"text-align: left\"><b>Ray version:</b></td>\n",
       "        <td style=\"text-align: left\"><b>2.42.1</b></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "    <td style=\"text-align: left\"><b>Dashboard:</b></td>\n",
       "    <td style=\"text-align: left\"><b><a href=\"http://127.0.0.1:8265\" target=\"_blank\">http://127.0.0.1:8265</a></b></td>\n",
       "</tr>\n",
       "\n",
       "</table>\n",
       "\n",
       "    </div>\n",
       "</div>\n"
      ],
      "text/plain": [
       "RayContext(dashboard_url='127.0.0.1:8265', python_version='3.11.3', ray_version='2.42.1', ray_commit='c2e38f7b75be223c0c033986472daada8622d64f')"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[36m(download_network pid=492099)\u001b[0m extracting public_transport for sarajevo\n",
      "\u001b[36m(download_network pid=492102)\u001b[0m extracting drive for sarajevo\n",
      "\u001b[36m(download_network pid=492101)\u001b[0m extracting bike for sarajevo\n",
      "\u001b[36m(download_network pid=492103)\u001b[0m extracting walk for sarajevo\n",
      "\u001b[36m(download_network pid=492099)\u001b[0m finish processing\n",
      "\u001b[36m(download_network pid=492099)\u001b[0m Final number of edges: 704\n",
      "\u001b[36m(download_network pid=492099)\u001b[0m Final number of nodes: 482\n",
      "\u001b[36m(download_network pid=492102)\u001b[0m finish download\n",
      "\u001b[36m(download_network pid=492102)\u001b[0m Raw number of edges: 58322\n",
      "\u001b[36m(download_network pid=492102)\u001b[0m Raw number of nodes: 30944\n",
      "\u001b[36m(download_network pid=492102)\u001b[0m start rebuilding graph\n",
      "\u001b[36m(download_network pid=492102)\u001b[0m finish rebuilding graph\n",
      "\u001b[36m(download_network pid=492102)\u001b[0m start graph simplify\n",
      "\u001b[36m(download_network pid=492102)\u001b[0m finish graph simplify\n",
      "\u001b[36m(download_network pid=492101)\u001b[0m finish download\u001b[32m [repeated 2x across cluster]\u001b[0m\n",
      "\u001b[36m(download_network pid=492103)\u001b[0m Raw number of edges: 110954\u001b[32m [repeated 2x across cluster]\u001b[0m\n",
      "\u001b[36m(download_network pid=492103)\u001b[0m Raw number of nodes: 51324\u001b[32m [repeated 2x across cluster]\u001b[0m\n",
      "\u001b[36m(download_network pid=492102)\u001b[0m finish processing\n",
      "\u001b[36m(download_network pid=492102)\u001b[0m Final number of edges: 11421\n",
      "\u001b[36m(download_network pid=492102)\u001b[0m Final number of nodes: 5173\n",
      "\u001b[36m(download_network pid=492101)\u001b[0m start rebuilding graph\u001b[32m [repeated 2x across cluster]\u001b[0m\n",
      "\u001b[36m(download_network pid=492101)\u001b[0m finish rebuilding graph\u001b[32m [repeated 2x across cluster]\u001b[0m\n",
      "\u001b[36m(download_network pid=492101)\u001b[0m start graph simplify\u001b[32m [repeated 2x across cluster]\u001b[0m\n",
      "\u001b[36m(download_network pid=492101)\u001b[0m finish graph simplify\n",
      "\u001b[36m(download_network pid=492101)\u001b[0m finish processing\n",
      "\u001b[36m(download_network pid=492101)\u001b[0m Final number of edges: 22887\n",
      "\u001b[36m(download_network pid=492101)\u001b[0m Final number of nodes: 10202\n",
      "\u001b[36m(download_network pid=492103)\u001b[0m finish graph simplify\n",
      "\u001b[36m(download_network pid=492103)\u001b[0m edges: 18628\n",
      "\u001b[36m(download_network pid=492103)\u001b[0m Number of partitions: 1\n",
      "\u001b[36m(_pedestrian_remove_sidewalks pid=492101)\u001b[0m starting with 18628 edges\n",
      "\u001b[36m(download_network pid=492103)\u001b[0m new graph MultiGraph with 14456 nodes and 17703 edges\n",
      "\u001b[36m(download_network pid=492103)\u001b[0m finish processing\n",
      "\u001b[36m(download_network pid=492103)\u001b[0m Final number of edges: 17167\n",
      "\u001b[36m(download_network pid=492103)\u001b[0m Final number of nodes: 13794\n"
     ]
    }
   ],
   "source": [
    "# Change for your own data base path\n",
    "data_base_path = \"/home/geolab/Desktop/Research/data\"\n",
    "\n",
    "# The extractor instance\n",
    "extractor = net_extractor.NetworkExtractor()\n",
    "extractor.DATA_BASE_PATH = data_base_path\n",
    " \n",
    "# Custom OSMnx settings\n",
    "settings.default_crs = \"epsg:4326\"\n",
    "\n",
    "ray.shutdown()\n",
    "ray.init()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract information from the GHS dataset\n",
    "urban_centers: geopd.GeoDataFrame\n",
    "\n",
    "ghs_dataset_path = f\"{data_base_path}/GHS_URBAN_CENTERS/GHS_URBAN_SIMPLIFIED_fixed.gpkg\"\n",
    "urban_centers = geopd.read_file(\n",
    "    ghs_dataset_path,\n",
    "    layer='GHS_URBAN_SIMPLIFIED_fixed'    \n",
    ")\n",
    "urban_centers = urban_centers.rename(columns={\n",
    "    \"GC_POP_TOT_2025\": \"population\",\n",
    "    \"GC_UCA_KM2_2025\": \"area\",\n",
    "    \"GC_DEV_USR_2025\": \"continent\",\n",
    "    \"GC_UCN_MAI_2025\": \"name\",\n",
    "    \"GC_CNT_GAD_2025\": \"country\"\n",
    "})\n",
    "transform = pyproj.Transformer.from_crs(\"ESRI:54009\", \"EPSG:4326\", always_xy=True).transform"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "capital_cities = f\"{data_base_path}/capital_cities.csv\"\n",
    "capital_df = pd.read_csv(capital_cities, delimiter=\",\", header=None)\n",
    "countries = list(capital_df[0])\n",
    "city_names = list(capital_df[1])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "capitals = pd.DataFrame()\n",
    "for i in range(len(countries)):\n",
    "    capital_city = urban_centers.loc[\n",
    "        (urban_centers[\"name\"] == city_names[i]) &\n",
    "        (urban_centers[\"country\"] == countries[i])\n",
    "    ]\n",
    "    capitals = pd.concat([capitals, capital_city])\n",
    "\n",
    "capitals = capitals.reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Kabul',\n",
       " 'Tirana',\n",
       " 'Algiers',\n",
       " 'Luanda',\n",
       " 'Buenos Aires',\n",
       " 'Yerevan',\n",
       " 'Sydney',\n",
       " 'Vienna',\n",
       " 'Baku',\n",
       " 'Nassau',\n",
       " 'Manama',\n",
       " 'Dhaka',\n",
       " 'Bridgetown',\n",
       " 'Minsk',\n",
       " 'Brussels',\n",
       " 'Belize City',\n",
       " 'Porto-Novo',\n",
       " 'Sucre',\n",
       " 'Sarajevo',\n",
       " 'Gaborone',\n",
       " 'Brasilia',\n",
       " 'Bandar Seri Begawan',\n",
       " 'Sofia',\n",
       " 'Ouagadougou',\n",
       " 'Gitega',\n",
       " 'Phnom Penh',\n",
       " 'Yaoundé',\n",
       " 'Ottawa',\n",
       " 'Bangui',\n",
       " \"N'Djamena\",\n",
       " 'Santiago',\n",
       " 'Beijing',\n",
       " 'Bogota',\n",
       " 'Moroni',\n",
       " 'Kinshasa',\n",
       " 'Brazzaville',\n",
       " 'San José',\n",
       " 'Yamoussoukro',\n",
       " 'Zagreb',\n",
       " 'Havana',\n",
       " 'Prague',\n",
       " 'Copenhagen',\n",
       " 'Djibouti',\n",
       " 'Santo Domingo',\n",
       " 'Dili',\n",
       " 'Quito',\n",
       " 'Cairo',\n",
       " 'San Salvador',\n",
       " 'London',\n",
       " 'Malabo',\n",
       " 'Asmara',\n",
       " 'Tallinn',\n",
       " 'Manzini',\n",
       " 'Addis Ababa',\n",
       " 'Suva',\n",
       " 'Helsinki',\n",
       " 'Paris',\n",
       " 'Libreville',\n",
       " 'Serrekunda',\n",
       " 'Tbilisi',\n",
       " 'Berlin',\n",
       " 'Accra',\n",
       " 'Athens',\n",
       " 'Guatemala City',\n",
       " 'Coyah',\n",
       " 'Bissau',\n",
       " 'Georgetown',\n",
       " 'Port-au-Prince',\n",
       " 'Tegucigalpa',\n",
       " 'Budapest',\n",
       " 'Reykjavik',\n",
       " 'New Delhi',\n",
       " 'Jakarta',\n",
       " 'Tehran',\n",
       " 'Baghdad',\n",
       " 'Dublin',\n",
       " 'Jerusalem',\n",
       " 'Rome',\n",
       " 'Kingston',\n",
       " 'Tokyo',\n",
       " 'Amman',\n",
       " 'Astana',\n",
       " 'Nairobi',\n",
       " 'Pristina',\n",
       " 'Kuwait City',\n",
       " 'Bishkek',\n",
       " 'Vientiane',\n",
       " 'Riga',\n",
       " 'Beirut',\n",
       " 'Maseru',\n",
       " 'Monrovia',\n",
       " 'Tripoli',\n",
       " 'Vilnius',\n",
       " 'Luxembourg',\n",
       " 'Antananarivo',\n",
       " 'Lilongwe',\n",
       " 'Kuala Lumpur',\n",
       " 'Malé',\n",
       " 'Bamako',\n",
       " 'Valletta',\n",
       " 'Nouakchott',\n",
       " 'Port Louis',\n",
       " 'Mexico City',\n",
       " 'Chișinău',\n",
       " 'Ulaanbaatar',\n",
       " 'Podgorica',\n",
       " 'Rabat',\n",
       " 'Maputo',\n",
       " 'Pyinmana [Nay Pyi Taw]',\n",
       " 'Windhoek',\n",
       " 'Kathmandu',\n",
       " 'Amsterdam',\n",
       " 'Wellington',\n",
       " 'Managua',\n",
       " 'Niamey',\n",
       " 'Abuja',\n",
       " \"P'yŏngyang\",\n",
       " 'Skopje',\n",
       " 'Belfast',\n",
       " 'Oslo',\n",
       " 'Muscat',\n",
       " 'Islamabad',\n",
       " 'Panama City',\n",
       " 'Port Moresby',\n",
       " 'Asuncion',\n",
       " 'Lima',\n",
       " 'Manila',\n",
       " 'Warsaw',\n",
       " 'Lisbon',\n",
       " 'Doha',\n",
       " 'Bucharest',\n",
       " 'Moscow',\n",
       " 'Kigali',\n",
       " 'Apia',\n",
       " 'São Tomé',\n",
       " 'Riyadh',\n",
       " 'Edinburgh',\n",
       " 'Dakar',\n",
       " 'Belgrade',\n",
       " 'Freetown',\n",
       " 'Singapore',\n",
       " 'Bratislava',\n",
       " 'Ljubljana',\n",
       " 'Honiara',\n",
       " 'Mogadishu',\n",
       " 'Pretoria',\n",
       " 'Seoul',\n",
       " 'Juba',\n",
       " 'Madrid',\n",
       " 'Colombo [Sri Jayawardenepura Kotte]',\n",
       " 'Khartoum',\n",
       " 'Paramaribo',\n",
       " 'Stockholm',\n",
       " 'Bern',\n",
       " 'Damascus',\n",
       " 'Taipei',\n",
       " 'Dushanbe',\n",
       " 'Dodoma',\n",
       " 'Bangkok',\n",
       " 'Lomé',\n",
       " \"Nuku'alofa\",\n",
       " 'Port of Spain',\n",
       " 'Tunis',\n",
       " 'Ankara',\n",
       " 'Ashgabat',\n",
       " 'Kampala',\n",
       " 'Kyiv',\n",
       " 'Abu Dhabi',\n",
       " 'London',\n",
       " 'Washington',\n",
       " 'Montevideo',\n",
       " 'Tashkent',\n",
       " 'Port Vila',\n",
       " 'Caracas',\n",
       " 'Hanoi',\n",
       " 'Cardiff',\n",
       " \"Sana'a\",\n",
       " 'Lusaka',\n",
       " 'Harare']"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(capitals[\"name\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_470632/765246397.py:3: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "  city_name = cap[1][1].replace(\" \", \"_\").lower()\n",
      "/tmp/ipykernel_470632/765246397.py:7: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "  search_name = cap[1][1]\n",
      "/tmp/ipykernel_470632/765246397.py:11: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "  country = cap[1][3]\n"
     ]
    }
   ],
   "source": [
    "\n",
    "for cap in capitals.iterrows():\n",
    "    # the city name in lowercase and slug_case for creating the folder to store the graphs and shapefiles\n",
    "    city_name = cap[1][1].replace(\" \", \"_\").lower()\n",
    "    #city_name = \"buenos_aires\"\n",
    "\n",
    "    # The name to search the city in the GHS dataset. Capital case. Also used for the DEM.\n",
    "    search_name = cap[1][1]\n",
    "    #search_name = \"Buenos Aires\"\n",
    "\n",
    "    # The country in which the city is located for searching the GHS dataset. Capital case.\n",
    "    country = cap[1][3]\n",
    "    #country = \"Argentina\"\n",
    "\n",
    "    # extract info from GHS with search_name and country\n",
    "    city_info = urban_centers.loc[\n",
    "        (urban_centers[\"name\"] == search_name) &\n",
    "        (urban_centers[\"country\"] == country)\n",
    "    ]\n",
    "\n",
    "    if len(city_info) == 0:\n",
    "        print(f\"City {search_name}, {country} not found\")\n",
    "\n",
    "    else:\n",
    "        geom = city_info[\"geometry\"].values[0]\n",
    "        geom = ops.transform(transform, geom)\n",
    "\n",
    "    o_b = \"{\"\n",
    "    c_b = \"}\"\n",
    "    backslash = f\"\\\\\"\n",
    "    specials = \"áéíóú'șăŏã\"\n",
    "    replaces = \"aeiou_saoa\"\n",
    "\n",
    "    display_name = city_name\n",
    "    modified = False\n",
    "\n",
    "    if \"[\" in display_name or \"]\" in display_name:\n",
    "        modified  = True\n",
    "\n",
    "    for i in range(len(specials)):\n",
    "        if specials[i] in display_name:\n",
    "            modified  = True\n",
    "        display_name = display_name.replace(specials[i], replaces[i])\n",
    "        display_name = display_name.replace(\"[\", \"\")\n",
    "        display_name = display_name.replace(\"]\", \"\")\n",
    "\n",
    "    ee_str = f\"{o_b}'geometry': ee.Geometry.BBox{geom.bounds}, 'name': '{display_name}' {c_b},\"\n",
    "\n",
    "    # if modified:\n",
    "    #     print(ee_str)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID_UC_G0</th>\n",
       "      <th>name</th>\n",
       "      <th>GC_UCN_LIS_2025</th>\n",
       "      <th>country</th>\n",
       "      <th>GC_CNT_UNN_2025</th>\n",
       "      <th>area</th>\n",
       "      <th>population</th>\n",
       "      <th>GC_DEV_WIG_2025</th>\n",
       "      <th>continent</th>\n",
       "      <th>GC_PLS_SCR_2025</th>\n",
       "      <th>GC_UCB_YOB_2025</th>\n",
       "      <th>GC_UCB_YOD_2025</th>\n",
       "      <th>GC_UCM_CAP</th>\n",
       "      <th>geometry</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>426</td>\n",
       "      <td>Sarajevo</td>\n",
       "      <td>Sarajevo</td>\n",
       "      <td>Bosnia and Herzegovina</td>\n",
       "      <td>Bosnia and Herzegovina</td>\n",
       "      <td>61</td>\n",
       "      <td>235739.1356</td>\n",
       "      <td>Upper Middle</td>\n",
       "      <td>Europe</td>\n",
       "      <td>High</td>\n",
       "      <td>1975</td>\n",
       "      <td>2030</td>\n",
       "      <td>1</td>\n",
       "      <td>MULTIPOLYGON (((1503001 5219001, 1506000 52190...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    ID_UC_G0      name GC_UCN_LIS_2025                 country  \\\n",
       "18       426  Sarajevo        Sarajevo  Bosnia and Herzegovina   \n",
       "\n",
       "           GC_CNT_UNN_2025  area   population GC_DEV_WIG_2025 continent  \\\n",
       "18  Bosnia and Herzegovina    61  235739.1356    Upper Middle    Europe   \n",
       "\n",
       "   GC_PLS_SCR_2025  GC_UCB_YOB_2025  GC_UCB_YOD_2025  GC_UCM_CAP  \\\n",
       "18            High             1975             2030           1   \n",
       "\n",
       "                                             geometry  \n",
       "18  MULTIPOLYGON (((1503001 5219001, 1506000 52190...  "
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "capitals = capitals.loc[capitals[\"name\"] == \"Sarajevo\"]\n",
    "capitals"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_470632/2370258519.py:4: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "  city_name = cap[1][1].replace(\" \", \"_\").lower()\n",
      "/tmp/ipykernel_470632/2370258519.py:8: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "  search_name = cap[1][1]\n",
      "/tmp/ipykernel_470632/2370258519.py:12: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "  country = cap[1][3]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(18.290715321625935, 43.80402003552376, 18.44633646029667, 43.895543123178676)\n",
      "u\n",
      "v\n",
      "key\n",
      "geometry\n",
      "length\n",
      "name\n",
      "highway\n",
      "bearing\n",
      "foot\n",
      "footway\n",
      "speed_kph\n",
      "travel_time\n",
      "grade\n",
      "grade_abs\n",
      "access\n",
      "sidewalk\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/geolab/Desktop/Research/notebooks/modules/network_extractor.py:288: UserWarning: Column names longer than 10 characters will be truncated when saved to ESRI Shapefile.\n",
      "  nodes.to_file(f\"{self.DATA_BASE_PATH}/{path}_nodes.shp\", encoding='utf-8')\n",
      "/home/geolab/Desktop/Research/.venv/lib/python3.11/site-packages/pyogrio/raw.py:723: RuntimeWarning: Normalized/laundered field name: 'street_count' to 'street_cou'\n",
      "  ogr_write(\n",
      "/home/geolab/Desktop/Research/notebooks/modules/network_extractor.py:290: UserWarning: Column names longer than 10 characters will be truncated when saved to ESRI Shapefile.\n",
      "  edges.to_file(f\"{self.DATA_BASE_PATH}/{path}_edges.shp\", encoding='utf-8')\n",
      "/home/geolab/Desktop/Research/.venv/lib/python3.11/site-packages/pyogrio/raw.py:723: RuntimeWarning: Normalized/laundered field name: 'travel_time' to 'travel_tim'\n",
      "  ogr_write(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "u\n",
      "v\n",
      "key\n",
      "geometry\n",
      "length\n",
      "name\n",
      "highway\n",
      "lanes\n",
      "speed_kph\n",
      "travel_time\n",
      "bearing\n",
      "grade\n",
      "grade_abs\n",
      "cycleway\n",
      "access\n",
      "bicycle\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/geolab/Desktop/Research/notebooks/modules/network_extractor.py:288: UserWarning: Column names longer than 10 characters will be truncated when saved to ESRI Shapefile.\n",
      "  nodes.to_file(f\"{self.DATA_BASE_PATH}/{path}_nodes.shp\", encoding='utf-8')\n",
      "/home/geolab/Desktop/Research/.venv/lib/python3.11/site-packages/pyogrio/raw.py:723: RuntimeWarning: Normalized/laundered field name: 'street_count' to 'street_cou'\n",
      "  ogr_write(\n",
      "/home/geolab/Desktop/Research/notebooks/modules/network_extractor.py:290: UserWarning: Column names longer than 10 characters will be truncated when saved to ESRI Shapefile.\n",
      "  edges.to_file(f\"{self.DATA_BASE_PATH}/{path}_edges.shp\", encoding='utf-8')\n",
      "/home/geolab/Desktop/Research/.venv/lib/python3.11/site-packages/pyogrio/raw.py:723: RuntimeWarning: Normalized/laundered field name: 'travel_time' to 'travel_tim'\n",
      "  ogr_write(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "u\n",
      "v\n",
      "key\n",
      "geometry\n",
      "length\n",
      "oneway\n",
      "lanes\n",
      "name\n",
      "highway\n",
      "maxspeed\n",
      "bearing\n",
      "speed_kph\n",
      "travel_time\n",
      "grade\n",
      "grade_abs\n",
      "ref\n",
      "junction\n",
      "width\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/geolab/Desktop/Research/notebooks/modules/network_extractor.py:288: UserWarning: Column names longer than 10 characters will be truncated when saved to ESRI Shapefile.\n",
      "  nodes.to_file(f\"{self.DATA_BASE_PATH}/{path}_nodes.shp\", encoding='utf-8')\n",
      "/home/geolab/Desktop/Research/.venv/lib/python3.11/site-packages/pyogrio/raw.py:723: RuntimeWarning: Normalized/laundered field name: 'street_count' to 'street_cou'\n",
      "  ogr_write(\n",
      "/home/geolab/Desktop/Research/notebooks/modules/network_extractor.py:290: UserWarning: Column names longer than 10 characters will be truncated when saved to ESRI Shapefile.\n",
      "  edges.to_file(f\"{self.DATA_BASE_PATH}/{path}_edges.shp\", encoding='utf-8')\n",
      "/home/geolab/Desktop/Research/.venv/lib/python3.11/site-packages/pyogrio/raw.py:723: RuntimeWarning: Normalized/laundered field name: 'travel_time' to 'travel_tim'\n",
      "  ogr_write(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "u\n",
      "v\n",
      "key\n",
      "geometry\n",
      "description\n",
      "from\n",
      "to\n",
      "operator\n",
      "network\n",
      "type\n",
      "route\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/geolab/Desktop/Research/notebooks/modules/network_extractor.py:290: UserWarning: Column names longer than 10 characters will be truncated when saved to ESRI Shapefile.\n",
      "  edges.to_file(f\"{self.DATA_BASE_PATH}/{path}_edges.shp\", encoding='utf-8')\n",
      "/home/geolab/Desktop/Research/.venv/lib/python3.11/site-packages/pyogrio/raw.py:723: RuntimeWarning: Normalized/laundered field name: 'description' to 'descriptio'\n",
      "  ogr_write(\n"
     ]
    }
   ],
   "source": [
    "errors = []\n",
    "for cap in capitals.iterrows():\n",
    "    # the city name in lowercase and slug_case for creating the folder to store the graphs and shapefiles\n",
    "    city_name = cap[1][1].replace(\" \", \"_\").lower()\n",
    "    #city_name = \"buenos_aires\"\n",
    "\n",
    "    # The name to search the city in the GHS dataset. Capital case. Also used for the DEM.\n",
    "    search_name = cap[1][1]\n",
    "    #search_name = \"Buenos Aires\"\n",
    "\n",
    "    # The country in which the city is located for searching the GHS dataset. Capital case.\n",
    "    country = cap[1][3]\n",
    "    #country = \"Argentina\"\n",
    "\n",
    "    o_b = \"{\"\n",
    "    c_b = \"}\"\n",
    "    backslash = f\"\\\\\"\n",
    "    specials = \"áéíóú'șăŏã\"\n",
    "    replaces = \"aeiou_saoa\"\n",
    "\n",
    "    display_name = city_name\n",
    "    modified = False\n",
    "\n",
    "    if \"[\" in display_name or \"]\" in display_name:\n",
    "        modified  = True\n",
    "\n",
    "    for i in range(len(specials)):\n",
    "        if specials[i] in display_name:\n",
    "            modified  = True\n",
    "        display_name = display_name.replace(specials[i], replaces[i])\n",
    "        display_name = display_name.replace(\"[\", \"\")\n",
    "        display_name = display_name.replace(\"]\", \"\")\n",
    "\n",
    "    city_name = display_name\n",
    "\n",
    "    try:\n",
    "        # extract info from GHS with search_name and country\n",
    "        city_info = urban_centers.loc[\n",
    "            (urban_centers[\"name\"] == search_name) &\n",
    "            (urban_centers[\"country\"] == country)\n",
    "        ]\n",
    "\n",
    "        if len(city_info) == 0:\n",
    "            print(f\"City {search_name}, {country} not found\")\n",
    "\n",
    "        else:\n",
    "            geom = city_info[\"geometry\"].values[0]\n",
    "            geom = ops.transform(transform, geom)\n",
    "\n",
    "        print(geom.bounds)\n",
    "\n",
    "        # the geometry from which to extract the network. Is given by the GHS dataset.\n",
    "        geometry = geom\n",
    "\n",
    "        # Variables for the elimination of duplicate pedestrian/driving streets\n",
    "        dist_threshold = 20\n",
    "        slope_threshold = 15\n",
    "\n",
    "        # Assessment = False, so duplicate pedestrian/driving streets will be eliminated.\n",
    "        assess = False\n",
    "\n",
    "        # Create graph and shapefile folders if they do not exist\n",
    "        Path(f\"{data_base_path}/{city_name}/graph\").mkdir(parents=True, exist_ok=True) # graphml folder\n",
    "        Path(f\"{data_base_path}/{city_name}/shp\").mkdir(parents=True, exist_ok=True) # shapefiles folder\n",
    "\n",
    "        # Process the 4 networks in parallel using the paralellized download_network function.\n",
    "        g_promises = []\n",
    "\n",
    "        # the pedestrian network\n",
    "        g_promises.append(extractor.download_network.remote(\n",
    "            extractor,\n",
    "            \"walk\", \n",
    "            geometry, \n",
    "            city_name, \n",
    "            assessment=assess, \n",
    "            dist_threshold=dist_threshold, \n",
    "            slope_threshold=slope_threshold,\n",
    "            add_elevation=True,\n",
    "        ))\n",
    "\n",
    "        # the cycling network\n",
    "        g_promises.append(extractor.download_network.remote(\n",
    "            extractor,\n",
    "            \"bike\", \n",
    "            geometry, \n",
    "            city_name, \n",
    "            assessment=assess, \n",
    "            add_elevation=True,\n",
    "        ))\n",
    "\n",
    "        # the driving network\n",
    "        g_promises.append(extractor.download_network.remote(\n",
    "            extractor,\n",
    "            \"drive\", \n",
    "            geometry, \n",
    "            city_name, \n",
    "            assessment=assess, \n",
    "            add_elevation=True,\n",
    "        ))\n",
    "\n",
    "        # the public transport network\n",
    "        g_promises.append(extractor.download_network.remote(\n",
    "            extractor,\n",
    "            \"public_transport\", \n",
    "            geometry, \n",
    "            city_name, \n",
    "            assessment=assess, \n",
    "            add_elevation=False,\n",
    "        ))\n",
    "\n",
    "        [g_walk, g_bike, g_drive, g_public] = ray.get(g_promises)\n",
    "\n",
    "        # Save graphs\n",
    "        extractor.save_as_graph(g_walk, f'{city_name}/graph/walk_{city_name}')\n",
    "        extractor.save_as_graph(g_bike, f'{city_name}/graph/bike_{city_name}')\n",
    "        extractor.save_as_graph(g_drive, f'{city_name}/graph/drive_{city_name}')\n",
    "        extractor.save_as_graph(g_public, f'{city_name}/graph/public_{city_name}')\n",
    "\n",
    "        # Save shapefiles\n",
    "        extractor.save_as_shp(g_walk, f'{city_name}/shp/walk_{city_name}')\n",
    "        extractor.save_as_shp(g_bike, f'{city_name}/shp/bike_{city_name}')\n",
    "        extractor.save_as_shp(g_drive, f'{city_name}/shp/drive_{city_name}')\n",
    "        extractor.save_as_shp(g_public, f'{city_name}/shp/public_{city_name}')\n",
    "    except Exception as ex:\n",
    "        errors.append(city_name)\n",
    "        print(ex)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['kabul',\n",
       " 'tirana',\n",
       " 'algiers',\n",
       " 'luanda',\n",
       " 'buenos_aires',\n",
       " 'yerevan',\n",
       " 'sydney',\n",
       " 'vienna',\n",
       " 'baku',\n",
       " 'nassau',\n",
       " 'manama',\n",
       " 'dhaka',\n",
       " 'bridgetown',\n",
       " 'minsk',\n",
       " 'brussels',\n",
       " 'belize_city',\n",
       " 'porto-novo',\n",
       " 'sucre',\n",
       " 'sarajevo',\n",
       " 'gaborone']"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-04-24 18:20:08,728\tERROR worker.py:422 -- Unhandled error (suppress with 'RAY_IGNORE_UNHANDLED_ERRORS=1'): multiprocessing.pool.RemoteTraceback: \n",
      "  File \"rasterio/_base.pyx\", line 310, in rasterio._base.DatasetBase.__init__\n",
      "  File \"rasterio/_base.pyx\", line 221, in rasterio._base.open_dataset\n",
      "  File \"rasterio/_err.pyx\", line 359, in rasterio._err.exc_wrap_pointer\n",
      "rasterio._err.CPLE_OpenFailedError: /home/geolab/Desktop/Research/data/DEM/Brasilia_DEM.tif: No such file or directory\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "\u001b[36mray::download_network()\u001b[39m (pid=399502, ip=10.78.3.89)\n",
      "  File \"/home/geolab/.pyenv/versions/3.11.3/lib/python3.11/multiprocessing/pool.py\", line 125, in worker\n",
      "    result = (True, func(*args, **kwds))\n",
      "                    ^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/geolab/.pyenv/versions/3.11.3/lib/python3.11/multiprocessing/pool.py\", line 51, in starmapstar\n",
      "    return list(itertools.starmap(args[0], args[1]))\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/geolab/Desktop/Research/.venv/lib/python3.11/site-packages/osmnx/elevation.py\", line 105, in _query_raster\n",
      "    with rasterio.open(filepath) as raster:\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/geolab/Desktop/Research/.venv/lib/python3.11/site-packages/rasterio/env.py\", line 463, in wrapper\n",
      "    return f(*args, **kwds)\n",
      "           ^^^^^^^^^^^^^^^^\n",
      "  File \"/home/geolab/Desktop/Research/.venv/lib/python3.11/site-packages/rasterio/__init__.py\", line 356, in open\n",
      "    dataset = DatasetReader(path, driver=driver, sharing=sharing, **kwargs)\n",
      "              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"rasterio/_base.pyx\", line 312, in rasterio._base.DatasetBase.__init__\n",
      "rasterio.errors.RasterioIOError: /home/geolab/Desktop/Research/data/DEM/Brasilia_DEM.tif: No such file or directory\n",
      "\"\"\"\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "\u001b[36mray::download_network()\u001b[39m (pid=399502, ip=10.78.3.89)\n",
      "  File \"/home/geolab/Desktop/Research/notebooks/modules/network_extractor.py\", line 393, in download_network\n",
      "    ox.add_node_elevations_raster(g, DEM_location, cpus=4)\n",
      "  File \"/home/geolab/Desktop/Research/.venv/lib/python3.11/site-packages/osmnx/elevation.py\", line 199, in add_node_elevations_raster\n",
      "    results = pool.starmap_async(_query_raster, args).get()\n",
      "              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/geolab/.pyenv/versions/3.11.3/lib/python3.11/multiprocessing/pool.py\", line 774, in get\n",
      "    raise self._value\n",
      "rasterio.errors.RasterioIOError: /home/geolab/Desktop/Research/data/DEM/Brasilia_DEM.tif: No such file or directory\n",
      "2025-04-24 18:20:19,873\tERROR worker.py:422 -- Unhandled error (suppress with 'RAY_IGNORE_UNHANDLED_ERRORS=1'): multiprocessing.pool.RemoteTraceback: \n",
      "  File \"rasterio/_base.pyx\", line 310, in rasterio._base.DatasetBase.__init__\n",
      "  File \"rasterio/_base.pyx\", line 221, in rasterio._base.open_dataset\n",
      "  File \"rasterio/_err.pyx\", line 359, in rasterio._err.exc_wrap_pointer\n",
      "rasterio._err.CPLE_OpenFailedError: /home/geolab/Desktop/Research/data/DEM/Minsk_DEM.tif: No such file or directory\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "\u001b[36mray::download_network()\u001b[39m (pid=396439, ip=10.78.3.89)\n",
      "  File \"/home/geolab/.pyenv/versions/3.11.3/lib/python3.11/multiprocessing/pool.py\", line 125, in worker\n",
      "    result = (True, func(*args, **kwds))\n",
      "                    ^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/geolab/.pyenv/versions/3.11.3/lib/python3.11/multiprocessing/pool.py\", line 51, in starmapstar\n",
      "    return list(itertools.starmap(args[0], args[1]))\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/geolab/Desktop/Research/.venv/lib/python3.11/site-packages/osmnx/elevation.py\", line 105, in _query_raster\n",
      "    with rasterio.open(filepath) as raster:\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/geolab/Desktop/Research/.venv/lib/python3.11/site-packages/rasterio/env.py\", line 463, in wrapper\n",
      "    return f(*args, **kwds)\n",
      "           ^^^^^^^^^^^^^^^^\n",
      "  File \"/home/geolab/Desktop/Research/.venv/lib/python3.11/site-packages/rasterio/__init__.py\", line 356, in open\n",
      "    dataset = DatasetReader(path, driver=driver, sharing=sharing, **kwargs)\n",
      "              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"rasterio/_base.pyx\", line 312, in rasterio._base.DatasetBase.__init__\n",
      "rasterio.errors.RasterioIOError: /home/geolab/Desktop/Research/data/DEM/Minsk_DEM.tif: No such file or directory\n",
      "\"\"\"\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "\u001b[36mray::download_network()\u001b[39m (pid=396439, ip=10.78.3.89)\n",
      "  File \"/home/geolab/Desktop/Research/notebooks/modules/network_extractor.py\", line 393, in download_network\n",
      "    ox.add_node_elevations_raster(g, DEM_location, cpus=4)\n",
      "  File \"/home/geolab/Desktop/Research/.venv/lib/python3.11/site-packages/osmnx/elevation.py\", line 199, in add_node_elevations_raster\n",
      "    results = pool.starmap_async(_query_raster, args).get()\n",
      "              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/geolab/.pyenv/versions/3.11.3/lib/python3.11/multiprocessing/pool.py\", line 774, in get\n",
      "    raise self._value\n",
      "rasterio.errors.RasterioIOError: /home/geolab/Desktop/Research/data/DEM/Minsk_DEM.tif: No such file or directory\n",
      "2025-04-24 18:20:30,733\tERROR worker.py:422 -- Unhandled error (suppress with 'RAY_IGNORE_UNHANDLED_ERRORS=1'): multiprocessing.pool.RemoteTraceback: \n",
      "  File \"rasterio/_base.pyx\", line 310, in rasterio._base.DatasetBase.__init__\n",
      "  File \"rasterio/_base.pyx\", line 221, in rasterio._base.open_dataset\n",
      "  File \"rasterio/_err.pyx\", line 359, in rasterio._err.exc_wrap_pointer\n",
      "rasterio._err.CPLE_OpenFailedError: /home/geolab/Desktop/Research/data/DEM/Brasilia_DEM.tif: No such file or directory\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "\u001b[36mray::download_network()\u001b[39m (pid=396046, ip=10.78.3.89)\n",
      "  File \"/home/geolab/.pyenv/versions/3.11.3/lib/python3.11/multiprocessing/pool.py\", line 125, in worker\n",
      "    result = (True, func(*args, **kwds))\n",
      "                    ^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/geolab/.pyenv/versions/3.11.3/lib/python3.11/multiprocessing/pool.py\", line 51, in starmapstar\n",
      "    return list(itertools.starmap(args[0], args[1]))\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/geolab/Desktop/Research/.venv/lib/python3.11/site-packages/osmnx/elevation.py\", line 105, in _query_raster\n",
      "    with rasterio.open(filepath) as raster:\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/geolab/Desktop/Research/.venv/lib/python3.11/site-packages/rasterio/env.py\", line 463, in wrapper\n",
      "    return f(*args, **kwds)\n",
      "           ^^^^^^^^^^^^^^^^\n",
      "  File \"/home/geolab/Desktop/Research/.venv/lib/python3.11/site-packages/rasterio/__init__.py\", line 356, in open\n",
      "    dataset = DatasetReader(path, driver=driver, sharing=sharing, **kwargs)\n",
      "              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"rasterio/_base.pyx\", line 312, in rasterio._base.DatasetBase.__init__\n",
      "rasterio.errors.RasterioIOError: /home/geolab/Desktop/Research/data/DEM/Brasilia_DEM.tif: No such file or directory\n",
      "\"\"\"\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "\u001b[36mray::download_network()\u001b[39m (pid=396046, ip=10.78.3.89)\n",
      "  File \"/home/geolab/Desktop/Research/notebooks/modules/network_extractor.py\", line 393, in download_network\n",
      "    ox.add_node_elevations_raster(g, DEM_location, cpus=4)\n",
      "  File \"/home/geolab/Desktop/Research/.venv/lib/python3.11/site-packages/osmnx/elevation.py\", line 199, in add_node_elevations_raster\n",
      "    results = pool.starmap_async(_query_raster, args).get()\n",
      "              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/geolab/.pyenv/versions/3.11.3/lib/python3.11/multiprocessing/pool.py\", line 774, in get\n",
      "    raise self._value\n",
      "rasterio.errors.RasterioIOError: /home/geolab/Desktop/Research/data/DEM/Brasilia_DEM.tif: No such file or directory\n",
      "2025-04-24 18:20:33,031\tERROR worker.py:422 -- Unhandled error (suppress with 'RAY_IGNORE_UNHANDLED_ERRORS=1'): multiprocessing.pool.RemoteTraceback: \n",
      "  File \"rasterio/_base.pyx\", line 310, in rasterio._base.DatasetBase.__init__\n",
      "  File \"rasterio/_base.pyx\", line 221, in rasterio._base.open_dataset\n",
      "  File \"rasterio/_err.pyx\", line 359, in rasterio._err.exc_wrap_pointer\n",
      "rasterio._err.CPLE_OpenFailedError: /home/geolab/Desktop/Research/data/DEM/Brussels_DEM.tif: No such file or directory\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "\u001b[36mray::download_network()\u001b[39m (pid=394063, ip=10.78.3.89)\n",
      "  File \"/home/geolab/.pyenv/versions/3.11.3/lib/python3.11/multiprocessing/pool.py\", line 125, in worker\n",
      "    result = (True, func(*args, **kwds))\n",
      "                    ^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/geolab/.pyenv/versions/3.11.3/lib/python3.11/multiprocessing/pool.py\", line 51, in starmapstar\n",
      "    return list(itertools.starmap(args[0], args[1]))\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/geolab/Desktop/Research/.venv/lib/python3.11/site-packages/osmnx/elevation.py\", line 105, in _query_raster\n",
      "    with rasterio.open(filepath) as raster:\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/geolab/Desktop/Research/.venv/lib/python3.11/site-packages/rasterio/env.py\", line 463, in wrapper\n",
      "    return f(*args, **kwds)\n",
      "           ^^^^^^^^^^^^^^^^\n",
      "  File \"/home/geolab/Desktop/Research/.venv/lib/python3.11/site-packages/rasterio/__init__.py\", line 356, in open\n",
      "    dataset = DatasetReader(path, driver=driver, sharing=sharing, **kwargs)\n",
      "              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"rasterio/_base.pyx\", line 312, in rasterio._base.DatasetBase.__init__\n",
      "rasterio.errors.RasterioIOError: /home/geolab/Desktop/Research/data/DEM/Brussels_DEM.tif: No such file or directory\n",
      "\"\"\"\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "\u001b[36mray::download_network()\u001b[39m (pid=394063, ip=10.78.3.89)\n",
      "  File \"/home/geolab/Desktop/Research/notebooks/modules/network_extractor.py\", line 393, in download_network\n",
      "    ox.add_node_elevations_raster(g, DEM_location, cpus=4)\n",
      "  File \"/home/geolab/Desktop/Research/.venv/lib/python3.11/site-packages/osmnx/elevation.py\", line 199, in add_node_elevations_raster\n",
      "    results = pool.starmap_async(_query_raster, args).get()\n",
      "              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/geolab/.pyenv/versions/3.11.3/lib/python3.11/multiprocessing/pool.py\", line 774, in get\n",
      "    raise self._value\n",
      "rasterio.errors.RasterioIOError: /home/geolab/Desktop/Research/data/DEM/Brussels_DEM.tif: No such file or directory\n",
      "2025-04-24 18:21:03,740\tERROR worker.py:422 -- Unhandled error (suppress with 'RAY_IGNORE_UNHANDLED_ERRORS=1'): multiprocessing.pool.RemoteTraceback: \n",
      "  File \"rasterio/_base.pyx\", line 310, in rasterio._base.DatasetBase.__init__\n",
      "  File \"rasterio/_base.pyx\", line 221, in rasterio._base.open_dataset\n",
      "  File \"rasterio/_err.pyx\", line 359, in rasterio._err.exc_wrap_pointer\n",
      "rasterio._err.CPLE_OpenFailedError: /home/geolab/Desktop/Research/data/DEM/Brasilia_DEM.tif: No such file or directory\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "\u001b[36mray::download_network()\u001b[39m (pid=399586, ip=10.78.3.89)\n",
      "  File \"/home/geolab/.pyenv/versions/3.11.3/lib/python3.11/multiprocessing/pool.py\", line 125, in worker\n",
      "    result = (True, func(*args, **kwds))\n",
      "                    ^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/geolab/.pyenv/versions/3.11.3/lib/python3.11/multiprocessing/pool.py\", line 51, in starmapstar\n",
      "    return list(itertools.starmap(args[0], args[1]))\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/geolab/Desktop/Research/.venv/lib/python3.11/site-packages/osmnx/elevation.py\", line 105, in _query_raster\n",
      "    with rasterio.open(filepath) as raster:\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/geolab/Desktop/Research/.venv/lib/python3.11/site-packages/rasterio/env.py\", line 463, in wrapper\n",
      "    return f(*args, **kwds)\n",
      "           ^^^^^^^^^^^^^^^^\n",
      "  File \"/home/geolab/Desktop/Research/.venv/lib/python3.11/site-packages/rasterio/__init__.py\", line 356, in open\n",
      "    dataset = DatasetReader(path, driver=driver, sharing=sharing, **kwargs)\n",
      "              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"rasterio/_base.pyx\", line 312, in rasterio._base.DatasetBase.__init__\n",
      "rasterio.errors.RasterioIOError: /home/geolab/Desktop/Research/data/DEM/Brasilia_DEM.tif: No such file or directory\n",
      "\"\"\"\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "\u001b[36mray::download_network()\u001b[39m (pid=399586, ip=10.78.3.89)\n",
      "  File \"/home/geolab/Desktop/Research/notebooks/modules/network_extractor.py\", line 393, in download_network\n",
      "    ox.add_node_elevations_raster(g, DEM_location, cpus=4)\n",
      "  File \"/home/geolab/Desktop/Research/.venv/lib/python3.11/site-packages/osmnx/elevation.py\", line 199, in add_node_elevations_raster\n",
      "    results = pool.starmap_async(_query_raster, args).get()\n",
      "              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/geolab/.pyenv/versions/3.11.3/lib/python3.11/multiprocessing/pool.py\", line 774, in get\n",
      "    raise self._value\n",
      "rasterio.errors.RasterioIOError: /home/geolab/Desktop/Research/data/DEM/Brasilia_DEM.tif: No such file or directory\n"
     ]
    }
   ],
   "source": [
    "errors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "# for typing help\n",
    "g_walk: nx.MultiDiGraph | nx.MultiGraph = g_walk\n",
    "\n",
    "to_remove = []\n",
    "for node in g_walk.nodes():\n",
    "    for node_to in g_walk[node]:\n",
    "        if len(list(g_walk[node][node_to].keys())) > 1:\n",
    "            min_parallel = 99999999\n",
    "            min_parallel_key = -1\n",
    "            for parallel_edge_key in g_walk[node][node_to].keys():\n",
    "                data = g_walk.get_edge_data(node, node_to, parallel_edge_key, default=None)\n",
    "                if data[\"length\"] > min_parallel:\n",
    "                    min_parallel = data[\"length\"]\n",
    "                    min_parallel_key = parallel_edge_key\n",
    "                \n",
    "            for parallel_edge_key in g_walk[node][node_to].keys():\n",
    "                if parallel_edge_key != min_parallel_key:\n",
    "                    to_remove.append([node, node_to, parallel_edge_key])\n",
    "            \n",
    "    \n",
    "    "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
